### ISSUE TITLE
Add support for a monorepo that requires everything remain on version lockstep

### ISSUE THREAD

**Issue by Noviny** [Noviny/changesets#7]
Required to support emotion

**Comment by mitchellhamilton** [comment]
We already discussed this but specifically, emotion needs to be able to keep _some_ packages in lockstep but not everything.

**Comment by Noviny** [comment]
👍 Thanks for clarifying! (It had fallen out of my brain)

**Comment by Noviny** [comment]
Q on this @mitchellhamilton, what is the expected changelog entry for this? Particularly if this is the only reason a package is being bumped?

**Comment by mitchellhamilton** [comment]
I don’t think a package should be bumped to keep all the versions the same unless it has changed or it needs to be updated because it's a dependent on a package that has changed so there won't be bumps just for this.

This is kind of how I think the logic would work. (this would happen in version-packages)

```jsx
let highestBumpType = getHighestBumpType(changesetsForlocksteppedPackages);

let currentLocksteppedVersion = require(rootPkgJsonPath).version;

let newLocksteppedVersion = bumpSemver(currentLocksteppedVersion, highestBumpType);

locksteppedPackagesWithChangesets.forEach(pkg => {
  versionPackage({ pkg, version: newLocksteppedVersion });
});

// then handle independently versioned packages like how they are right now...
```


The important thing to me is that when packages don’t get bumped in one release because they don't have changes and then later when those packages needs to be released, they need to be bumped to the lockstep version so essentially packages need to skip versions so they’re up to date with the other lockstep packages but if they don't have a changeset, they don't get a release.

I'm not totally sure about all of this though, I feel like this needs more exploration.

**Comment by Noviny** [comment]
So for practical example, if I have `packageA` and `packageB`, both on `1.0.0` with a lockStepVersion of `1.0.0`, I minor `packageA`, meaning `lockstepVersion` becomes `1.1.0`, and `packageA` is released at `1.1.0`, the latest version of `packageB` remains at `1.0.0`.

I then changeset patch `packageB`, the newLockstepVersion should be left at `1.1.0` and `packageB` is now published at `1.1.0`?

Or is it newLockstepVersion is pushed to `1.1.1` and `packageB` is now at `1.1.1`?

Also my assumption for lockstep is that you just know if you install two packages at the same version (react and react-dom at 16.8.0 for example) - either method of this doesn't technically create this symbiosis and surety.

**Comment by mitchellhamilton** [comment]
> Or is it newLockstepVersion is pushed to 1.1.1 and packageB is now at 1.1.1?

This one. Every time there's a release which contains lockstep packages, the lockstep version should increase.

> Also my assumption for lockstep is that you just know if you install two packages at the same version (react and react-dom at 16.8.0 for example) - either method of this doesn't technically create this symbiosis and surety.

I don't get this, could you elaborate?

**Comment by Noviny** [comment]
So, in my mind, the reason to do lockstep instead of just semver is that if you put the packages all on the same version, you know they all work together.

With the 1.1.1, you don't get that, because `packageB` doesn't have a `1.1.0`, so you either install:

```
packageA: 1.1.0,
packageB: 1.0.0
```

or

```
packageA: 1.1.0
packageB: 1.1.1
```

At which point I'm not sure what the goal of the lockstep is? What is gained by packageB getting neither 1.0.1 (just following semver) or 1.1.0 (more strict lockstep)

**Comment by mitchellhamilton** [comment]
> So, in my mind, the reason to do lockstep instead of just semver is that if you put the packages all on the same version, you know they all work together.

This is a really good point, the answer might be that a more strict lockstep is the approach to take.

> At which point I'm not sure what the goal of the lockstep is? What is gained by packageB getting neither 1.0.1 (just following semver) or 1.1.0 (more strict lockstep)

I feel like non-strict lockstep was nice with lerna's approach to versioning because it meant that you didn't have to think about versions for each individual packages but since changesets make it easier to specify updates and moves the job of specifying what kind of release something should be onto the person who is changing the package rather than the person who is doing the release so maybe the non-strict lockstep approach doesn't make sense with changesets so I'm starting to think that the more strict lockstep approach is better. Though something feels off about publishing new versions of packages that are the exact same as the previous version.

So I guess that brings us back to your original question about what to put in the changelog, I'm not really sure what the right answer is tbh. Maybe something like "Update to keep in lockstep with other packages"? but I feel like that doesn't sound great. I guess this is hard because the changelog is about changes and this would be a release that doesn't have any changes so it doesn't really make sense.

☝️ is kind of a reason why non strict lockstep could be good


(I wrote the above with the assumption that strict lockstep would be, on every release with lockstep packages, every one of those packages gets released but I think my assumption could be wrong?)

---
> So for practical example, if I have packageA and packageB, both on 1.0.0 with a lockStepVersion of 1.0.0, I minor packageA, meaning lockstepVersion becomes 1.1.0, and packageA is released at 1.1.0, the latest version of packageB remains at 1.0.0.
>
> I then changeset patch packageB, the newLockstepVersion should be left at 1.1.0 and packageB is now published at 1.1.0?

I just re-read this and it seems confusing, it would mean that a single version number which is intended to group multiple packages together would refer to multiple different releases? If I'm understanding what you mean by strict lockstep correctly then what is gained with strict lockstep 

I feel like what's nice about the "non strict lockstep" is that it makes it clear what's a new version because with my understanding of what you mean by strict lockstep, a consumer could have `@my-scope/a@1.1.0` and `@my-scope/b@1.0.0` and then `@my-scope/b` gets an update to `@my-scope/b@1.1.0`, so the consumer updates `@my-scope/b@1.0.0` to `@my-scope/b@1.1.0` but it seems confusing since the version numbers are so close and they are on the same scope so it seems like they were using an older version of `@my-scope/b` when a newer version existed when in reality, it didn't.

---

(apologies if this doesn't make sense, I'm trying to capture my thought process of this)
